src/Tasks/AssemblyDependency/ReferenceTable.cs

```diff
@@ -2690,65 +2686,68 @@ private ITaskItem SetItemMetadata(List<ITaskItem> relatedItems, List<ITaskItem>
-                {
-                    referenceItem.RemoveMetadata(ItemMetadataNames.winmdImplmentationFile);
-                }
-
-                // If the item originally did not have the ImageRuntime metadata then we do not want to get it from the set of primary source items
-                // since the ImageRuntime is something specific to the source item and not supposed to be propagated.
-                if (!hasImageRuntime)
-                {
-                    referenceItem.RemoveMetadata(ItemMetadataNames.imageRuntime);
-                }
-
-                // If the item originally did not have the WinMDFile metadata then we do not want to get it from the set of primary source items
-                // since the WinMDFile is something specific to the source item and not supposed to be propigated
-                if (!hasWinMDFile)
-                {
-                    referenceItem.RemoveMetadata(ItemMetadataNames.winMDFile);
-                }
             }
 
-            referenceItem.SetMetadata(ItemMetadataNames.version, reference.ReferenceVersion == null ? string.Empty : reference.ReferenceVersion.ToString());
+            // PERF: Order adds/removes after the initial copy to allow copy-on-write cloning between TaskItems.
+            // Overwrite any common metadata copied from source items.
+            IMetadataContainer referenceItemAsMetadataContainer = referenceItem;
+            referenceItemAsMetadataContainer.ImportMetadata(
+                EnumerateCommonMetadata(reference, assemblyName, _frameworkPaths, _installedAssemblies));
 
-            // Unset fusionName so we don't have to unset it later.
-            referenceItem.RemoveMetadata(ItemMetadataNames.fusionName);
+            // Some metadata should not be forwarded between the parent and child items.
+            // Set up a base item for related files. If we need to exclude any metadata, this will be a clone of the parent item.
+            TaskItem relatedItemBase = referenceItem;
 
             List<string> relatedFileExtensions = reference.GetRelatedFileExtensions();
             List<string> satellites = reference.GetSatelliteFiles();
             List<string> serializationAssemblyFiles = reference.GetSerializationAssemblyFiles();
             string[] scatterFiles = reference.GetScatterFiles();
-            Dictionary<string, string> nonForwardableMetadata = null;
-            if (relatedFileExtensions.Count > 0 || satellites.Count > 0 || serializationAssemblyFiles.Count > 0 || scatterFiles.Length > 0)
+            bool hasRelatedFiles = relatedFileExtensions.Count > 0 || satellites.Count > 0 || serializationAssemblyFiles.Count > 0 || scatterFiles.Length > 0;
+
+            // Remove metadata entries should not be copied to the output item or any related files.
+            // The specific metadata to remove depends on whether we found related files and the type of source item.
+            if (hasRelatedFiles)
```

Comment:
So a lot of the complexity here is just maintaining behavior, but after digging through commit history, I think it's safe to simplify this down further since there's a number of things that are just a result of code churn:

- `msbuildReferenceSourceTarget` is propogated to the WinMD implementation item - this happened in an unrelated perf refactor
- If no other related files (satellites, scatter files, ect.) are found, then the non-forwardable metadata *is* propogated to the WinMD implementation item - looks like an unintentional edge case due to refactoring
- Different metadata is removed depending on whether there's a primary source item - again seems only relevant for related items since we overwrite these at the end

Reply:
^^Done